File: apps/cli/src/commands/agent-services.ts
/**
 * `settlekit agent-services` — list and publish AI agent service listings.
 *
 *   list                 GET  /v1/agent-services
 *   publish <id>         POST /v1/agent-services/:id/publish
 *   metadata <id>        GET  /v1/agent-services/:id/metadata.json
 */
import type { Command } from "commander";
import { buildContext } from "../context.js";

interface AgentService extends Record<string, unknown> {
  id: string;
  name: string;
  price: string;
  network: string;
  status?: string;
}

export function registerAgentServices(program: Command): void {
  const agents = program.command("agent-services").description("List + publish AI agent services");

  agents
    .command("list")
    .description("List agent services")
    .action(async function (this: Command) {
      const ctx = buildContext(this);
      const rows = await ctx.client.get<AgentService[]>("/v1/agent-services");
      ctx.printList(rows, [
        { header: "ID", value: (a) => a.id },
        { header: "NAME", value: (a) => a.name },
        { header: "PRICE", value: (a) => a.price },
        { header: "NETWORK", value: (a) => a.network },
        { header: "STATUS", value: (a) => a.status },
      ]);
    });

  agents
    .command("publish <id>")
    .description("Publish an agent service")
    .action(async function (this: Command, id: string) {
      const ctx = buildContext(this);
      const service = await ctx.client.post<AgentService>(
        `/v1/agent-services/${encodeURIComponent(id)}/publish`,
      );
      ctx.printRecord(service);
    });

  agents
    .command("metadata <id>")
    .description("Fetch the agent-readable metadata.json for a service")
    .action(async function (this: Command, id: string) {
      const ctx = buildContext(this);
      const metadata = await ctx.client.get<Record<string, unknown>>(
        `/v1/agent-services/${encodeURIComponent(id)}/metadata.json`,
      );
      ctx.printRecord(metadata);
    });
}

File: sdks/go/agentservices.go
package settlekit

import (
	"context"
	"net/http"
)

// CreateAgentServiceInput is the body for CreateAgentService
// (POST /v1/agent-services).
type CreateAgentServiceInput struct {
	OrganizationID string         `json:"organizationId"`
	MerchantID     string         `json:"merchantId"`
	ProductID      string         `json:"productId,omitempty"`
	Name           string         `json:"name"`
	Description    string         `json:"description,omitempty"`
	Endpoint       string         `json:"endpoint"`
	Amount         string         `json:"amount"`
	Currency       string         `json:"currency,omitempty"`
	Network        string         `json:"network,omitempty"`
	Capabilities   []string       `json:"capabilities,omitempty"`
	Metadata       map[string]any `json:"metadata,omitempty"`
}

// CreateAgentService registers a machine-callable agent service.
// POST /v1/agent-services.
func (c *Client) CreateAgentService(ctx context.Context, in CreateAgentServiceInput) (*AgentService, error) {
	var out AgentService
	if err := c.do(ctx, http.MethodPost, "/v1/agent-services", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListAgentServices lists agent services, optionally filtered by organization id
// (pass "" for all). GET /v1/agent-services[?organizationId=].
func (c *Client) ListAgentServices(ctx context.Context, organizationID string) ([]AgentService, error) {
	q := newQuery()
	q.add("organizationId", organizationID)
	var out []AgentService
	if err := c.do(ctx, http.MethodGet, "/v1/agent-services"+q.encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// PublishAgentService publishes an agent service, making it discoverable and its
// metadata document public. POST /v1/agent-services/:id/publish.
func (c *Client) PublishAgentService(ctx context.Context, id string) (*AgentService, error) {
	var out AgentService
	if err := c.do(ctx, http.MethodPost, "/v1/agent-services/"+id+"/publish", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AgentServiceMetadata fetches the public metadata.json document for an agent
// service, describing how an agent invokes and pays for it.
// GET /v1/agent-services/:id/metadata.json.
func (c *Client) AgentServiceMetadata(ctx context.Context, id string) (*AgentServiceMetadata, error) {
	var out AgentServiceMetadata
	if err := c.do(ctx, http.MethodGet, "/v1/agent-services/"+id+"/metadata.json", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
